src/backend/sgx/attestation.rs

```diff
@@ -1,17 +1,77 @@
 // SPDX-License-Identifier: Apache-2.0
 
+// Credit to: https://github.com/fortanix/rust-sgx/tree/master/aesm-client
+// for examples of AESM Requests.
+
+use crate::protobuf::aesm_proto::{
+    Request, Request_InitQuoteRequest, Response, Response_InitQuoteResponse,
+};
 use crate::syscall::{SGX_DUMMY_QUOTE, SGX_DUMMY_TI, SGX_QUOTE_SIZE, SGX_TI_SIZE};
 
-use std::io::Error;
+use std::io::{Error, Read, Write};
+use std::mem::size_of;
+use std::os::unix::net::UnixStream;
 use std::slice::from_raw_parts_mut;
 
+use protobuf::{Message, ProtobufResult};
+
+const AESM_SOCKET: &str = "/var/run/aesmd/aesm.socket";
+
 /// Fills the Target Info of the QE into the output buffer specified and
 /// returns the number of bytes written.
-// TODO: Replace synthetic TargetInfo with real TargetInfo.
-fn get_ti(out_buf: &mut [u8]) -> usize {
+fn get_ti(out_buf: &mut [u8]) -> Result<usize, std::io::Error> {
     assert_eq!(out_buf.len(), SGX_TI_SIZE, "Invalid size of output buffer");
-    out_buf.copy_from_slice(&SGX_DUMMY_TI);
-    SGX_TI_SIZE
+
+    // If unable to connect to the AESM daemon, return dummy value
+    let mut stream = match UnixStream::connect(AESM_SOCKET) {
+        Ok(s) => s,
+        Err(_) => {
+            out_buf.copy_from_slice(&SGX_DUMMY_TI);
+            return Ok(SGX_TI_SIZE);
+        }
+    };
+
+    // Set an Init Quote Request
+    let mut req = Request::new();
+    let mut msg = Request_InitQuoteRequest::new();
+    msg.set_timeout(1_000_000);
+    req.set_initQuoteReq(msg);
+
+    // Set up Writer
+    let mut buf_wrtr = vec![0u8; size_of::<u32>()];
+    match req.write_to_writer(&mut buf_wrtr) {
+        Ok(_) => {}
+        Err(e) => eprintln!("Error getting TargetInfo: {:#?}", e),
+    }
+
+    let req_len = (buf_wrtr.len() - size_of::<u32>()) as u32;
+    (&mut buf_wrtr[0..size_of::<u32>()]).copy_from_slice(&req_len.to_le_bytes());
+
+    // Send Request to AESM daemon
+    stream.write_all(&buf_wrtr)?;
+    stream.flush()?;
+
+    // Receive Response
+    let mut res_len_bytes = [0u8; 4];
+    stream.read_exact(&mut res_len_bytes)?;
+    let res_len = u32::from_le_bytes(res_len_bytes);
+
+    let mut res_bytes = vec![0; res_len as usize];
+    stream.read_exact(&mut res_bytes)?;
+
+    // Parse Response and extract TargetInfo
+    let pb_msg: ProtobufResult<Response> = protobuf::parse_from_bytes(&res_bytes);
+    let res: Response_InitQuoteResponse = pb_msg.unwrap().take_initQuoteRes();
```

Comment:
I'm not familiar at all with protobuf types, so no worries if this doesn't "translate" well, but could this be squashed into something like this to avoid an unwrap:

```rust
let ti = protobuf::parse_from_bytes::<Response>(&res_bytes)
    .map(|msg| msg.take_initQuoteRes().get_targetInfo())?;
```